### PR TITLE
Openssl 1.1.0 fixes

### PR DESCRIPTION
**Pull request by jcard0na**
Fix authsae so it will work with openssl 1.1.0.

Tested by running integration test `test001.sh` against libcrypto 1.1.0 and 1.0.2.

### CODE REVIEW

File: sae.c [hunk 1]
```diff
@@ -76,6 +76,13 @@
 #define CN_Update(ctx,x,l) HMAC_Update((ctx),(x),(l))
 #define CN_Final(ctx,x) HMAC_Final((ctx), (x), &function_mdlen)
 
+/* Pre 1.1.0 compatibility macros */
+#ifndef OPENSSL_API_COMPAT
```
Comment (bcopeland) [inline sae.c]: where does this get set?

File: sae.c [hunk 2]
```diff
@@ -444,14 +458,14 @@ process_confirm (struct candidate *peer, struct ieee80211_mgmt_frame *frame, int
...
     offset = BN_num_bytes(peer->grp_def->prime) - BN_num_bytes(y);
     memset(tmp, 0, offset);
     BN_bn2bin(y, tmp + offset);
-    CN_Update(&ctx, tmp, BN_num_bytes(peer->grp_def->prime));
+    CN_Update(ctx, tmp, BN_num_bytes(peer->grp_def->prime));
 
-    CN_Final(&ctx, tmp);
-    HMAC_CTX_cleanup(&ctx);
+    CN_Final(ctx, tmp);
+    HMAC_CTX_free(ctx);
```
Comment (bcopeland) [inline sae.c]: I think this should go to out, otherwise there are leaks on the error paths

File: sae.c [hunk 3]
```diff
@@ -68,14 +68,21 @@
...
 
-#define CN_Init(ctx,x,l) HMAC_Init((ctx), (x), (l), EVP_sha256())
+#define CN_Init(ctx,x,l) HMAC_Init_ex((ctx), (x), (l), EVP_sha256(), NULL)
 #define CN_Update(ctx,x,l) HMAC_Update((ctx),(x),(l))
 #define CN_Final(ctx,x) HMAC_Final((ctx), (x), &function_mdlen)
 
+/* Pre 1.1.0 compatibility macros */
+#ifndef OPENSSL_API_COMPAT
+#define HMAC_CTX_new() calloc(sizeof(HMAC_CTX), 1)
+#define HMAC_CTX_free(ctx) { HMAC_CTX_cleanup(ctx); free(ctx); ctx = NULL; }
```
Comment (bcopeland) [inline sae.c]: not sure about openssl's semantics, but usually free() can be called on null which can make error paths a little easier to write, so I would suggest wrapping it all in `if (ctx) {` .  Also static function might be cleaner than macros.

File: sae.c [hunk 4]
```diff
@@ -483,9 +497,13 @@ confirm_to_peer (struct candidate *peer)
...
     size_t len = 0;
     BIGNUM *x, *y;
-    HMAC_CTX ctx;
+    HMAC_CTX *ctx;
     unsigned short send_conf;
     int offset;
+    ctx = HMAC_CTX_new();
+    if (ctx == NULL) {
+        return -1;
+    }
```
Comment (bcopeland) [inline sae.c]: This ctx will also leak if there are any errors, need to add a free in all the error paths or (probably cleaner) make a goto label at the end to cleanup the EC points x,y and the hmac context

File: sae.c [hunk 5]
```diff
@@ -596,7 +614,11 @@ process_commit (struct candidate *peer, struct ieee80211_mgmt_frame *frame, int
     int offset, itemsize, ret = 0;
     EC_POINT *K;
     unsigned char *ptr, *tmp, keyseed[SHA256_DIGEST_LENGTH], kckpmk[(SHA256_DIGEST_LENGTH * 2) * 8];
-    HMAC_CTX ctx;
+    HMAC_CTX *ctx;
+    ctx = HMAC_CTX_new();
+    if (ctx == NULL) {
+        return -1;
+    }
```
Comment (bcopeland) [inline sae.c]: same leaks here

File: sae.c [hunk 6]
```diff
@@ -2253,7 +2287,11 @@ JC: Commented out until we decide whether this is needed (in which case we must
 #endif
     TAILQ_INIT(&peers);
     TAILQ_INIT(&blacklist);
-    RAND_pseudo_bytes((unsigned char *)&token_generator, sizeof(unsigned long));
+    i = RAND_bytes((unsigned char *)&token_generator, sizeof(unsigned long));
+    if (i == 0) {
```
Comment (bcopeland) [inline sae.c]: what is this return value, number of bytes generated?  should we check that i == sizeof(unsigned long) instead?

File: sae.c (hunk 1 above)
Comment (jcard0na) [reply sae.c]: This is introduced in 1.1.0, defined `include/openssl/opensslconf.h`, and included in this module by `openssl/hmac.h`

File: sae.c (hunk 6 above)
Comment (jcard0na) [reply sae.c]: It's 1 on success, 0 on failure.  https://www.openssl.org/docs/man1.1.0/crypto/RAND_bytes.html

File: sae.c (hunk 3 above)
Comment (jcard0na) [reply sae.c]: will change

File: sae.c (hunk 2 above)
Comment (jcard0na) [reply sae.c]: will fix

File: sae.c (hunk 4 above)
Comment (jcard0na) [reply sae.c]: will fix

File: sae.c (hunk 5 above)
Comment (jcard0na) [reply sae.c]: will fix

File: sae.c (hunk 6 above)
Comment (bcopeland) [reply sae.c]: ah right, openssl has terrible ideas about return values... https://www.openssl.org/news/secadv/20090107.txt and I seem to remember others.

File: sae.c [hunk 7]
```diff
@@ -513,57 +545,54 @@ confirm_to_peer (struct candidate *peer)
...
-    CN_Update(&ctx, tmp, BN_num_bytes(peer->grp_def->order));
+    CN_Update(ctx, tmp, BN_num_bytes(peer->grp_def->order));
 
     if (!EC_POINT_get_affine_coordinates_GFp(peer->grp_def->group, peer->my_element, x, y, bnctx)) {
         sae_debug(SAE_DEBUG_ERR, "unable to get x,y of my element\n");
-        BN_free(x);
-        BN_free(y);
-        return -1;
+        ret = -1;
+        goto fail;
```
Comment (bcopeland) [inline sae.c]: heh nice label name (https://gotofail.com)

**Comment by bcopeland** [comment]
looks good, thanks

**Pull request merged**